### ISSUE TITLE
Defined boolean option not working in class that does not implement Runnable

### ISSUE THREAD

**Issue by elyograg** [remkop/picocli#2022]
I have added a "-v" option to enable debug logging.  My program ignores this option.
The only thing that I can think of which is different about the main command class compared to a command class where boolean options DO work is that the main command class does NOT implement Runnable.  I made this change because the docs said that was how to require a subcommand.

I will put the class file in a comment.

**Comment by elyograg** [comment]
[MainCommand.java.txt](https://github.com/remkop/picocli/files/11416527/MainCommand.java.txt)

**Comment by elyograg** [comment]
The -h option works, but none of the others do.

**Comment by remkop** [comment]
I am unable to view the content of the link you posted. Can you post the contents instead of a link please?

**Comment by remkop** [comment]
Also please specify how you are invoking the program.
I assume the `-v` option is specified in the top-level command, and you are invoking a subcommand? (Just guessing, haven't seen the code yet...) 
So, is the question: "how can subcommands detect which options were specified for their parent command"?
Riffing further on this theme (if I'm not wildly off track), one idea may be to make the `-v` option an [inherited option](https://picocli.info/#_inherited_options).

**Comment by remkop** [comment]
okay, I was able to see the code for MainCommand on my phone.

First, the `verbose` flag, and the `configFile` field, will not be populated until after the `CommandLine::execute` method is invoked.
That method will parse the user input, and execute the subcommand that the user specified.
From within the business logic of the subcommand, the `verbose` flag, and the `configFile` field can be accessed and will have the expected value from the user input.

You made the fields `static`, so the subcommands can just inspect `MainCommand.verbose` and `MainCommand.configFile` to access these values from the business logic of the subcommand.

Alternatively, if you want to do some validation (or switch on logging) _after parsing completed_ but _before executing any subcommand_, then one idea is to specify a [custom execution strategy](https://picocli.info/#_use_case_configure_log_level_with_a_global_option), which does your validation first, before doing its normal behaviour.

For example:

```java
class MainCommand {
    // ...
    public void static main(String... args) {
        MainCommand app = new MainCommand();
        new CommandLine(app)
                .setHelpFactory(createHelpFactory()) // your existing customization
                .setExecutionStrategy(app::executionStrategy) // <-- new
                .execute(args);
    }
    
    // do custom validation and enable logging here
    private int executionStrategy(ParseResult parseResult) {
        StaticStuff.setVerboseFlag(verbose);
        if (!StaticStuff.parseAndValidateConfig(configFile)) {
            throw new ParameterException(parseResult.commandSpec().commandLine(),
                    String.format("Invalid config file %s", configFile));
        }
        
        return new CommandLine.RunLast().execute(parseResult); // default execution strategy
    }
    
}

**Comment by elyograg** [comment]
I am running `./solr -v` ... a shell script wrapper.  The effective result of this is `java -jar /path/to/jar -v`

**Comment by remkop** [comment]
> I am running `./solr -v` ... a shell script wrapper. The effective result of this is `java -jar /path/to/jar -v`

I would expect picocli to respond to that with an error "Missing required subcommand" (because the MainCommand does not implement Runnable, see [this section](https://picocli.info/#_required_subcommands) in the manual for details). Is that what you are seeing?

**Comment by elyograg** [comment]
Thanks for the tip!  It's working.  To simplify things, I replaced FileNotFoundException with RuntimeException so I don't have to catch it or add "throws" to method signatures.

![image](https://user-images.githubusercontent.com/4984827/236722473-e1b0a7a6-2440-47da-a344-6c389d954ee8.png)

```
package org.apache.solr.cli;

import java.lang.invoke.MethodHandles;

import org.apache.solr.cli.commands.InfoCommand;
import org.apache.solr.cli.commands.StartCommand;
import org.apache.solr.cli.commands.StopCommand;
import org.apache.solr.cli.commands.ZkCommand;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import picocli.CommandLine;
import picocli.CommandLine.Command;
import picocli.CommandLine.Option;
import picocli.CommandLine.ParameterException;
import picocli.CommandLine.ParseResult;
import picocli.CommandLine.ScopeType;

/** The main command class. */
@Command(name = "solr", sortOptions = false, version = StaticStuff.SOLR_VERSION, scope = ScopeType.INHERIT, description = "A script that controls Solr and related functionality.", synopsisSubcommandLabel = "COMMAND", subcommands = {
    StartCommand.class, StopCommand.class, ZkCommand.class,
    InfoCommand.class }, footer = StaticStuff.USAGE_OPTION_SEPARATOR_TEXT)
public final class MainCommand {

  /** Logger object. */
  @SuppressWarnings("unused")
  private static final Logger log = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());

  /** Help option. */
  @Option(names = { "-h",
      "--help" }, arity = "0", usageHelp = true, scope = ScopeType.INHERIT, description = "Display this command usage.")
  private static boolean help;

  /** Configuration file option. */
  @Option(names = { "-config",
      "--config" }, arity = "1", scope = ScopeType.INHERIT, description = "The location of the cross-platform config file.  If not present, program will search predefined paths for the file.")
  private static String configFile = null;

  /** Debug option. */
  @Option(names = { "-v", "--verbose",
      "--debug" }, arity = "0", scope = ScopeType.INHERIT, description = "Log any available debug messages.  Some commands may not have any debug logging defined.")
  private static boolean verbose;

  /** Debug option. */
  @Option(names = { "-f",
      "--force" }, arity = "0", scope = ScopeType.INHERIT, description = "Force running as root, and as a side effect, allow binding to ports below 1024.")
  private static boolean force;

  /**
   * The main method.
   *
   * @param args command-line arguments
   */
  public static final void main(final String[] args) {
    final MainCommand app = new MainCommand();
    new CommandLine(app).setHelpFactory(StaticStuff.createLeftAlignedUsageHelp())
        .setExecutionStrategy(app::executionStrategy).execute(args);
  }

  /**
   * Do custom validation and enable logging here.
   * 
   * @param parseResult
   * @return
   */
  private int executionStrategy(final ParseResult parseResult) {
    StaticStuff.setVerboseFlag(verbose);
    StaticStuff.setForceFlag(force);
      if (!StaticStuff.parseAndValidateConfig(configFile)) {
        throw new ParameterException(parseResult.commandSpec().commandLine(),
            String.format("Something is amiss with the config."));
      }

    return new CommandLine.RunLast().execute(parseResult); // default execution strategy
  }
}
```

**Comment by elyograg** [comment]
> I would expect picocli to respond to that with an error "Missing required subcommand" (because no subcommand is specified, and the `MainCommand` does not implement Runnable, see [this section](https://picocli.info/#_required_subcommands) in the manual for details). Is that what you are seeing?

Yes, it does say the subcommand is missing.  Before, it did not show debug logs before giving that error, but since adding the execution strategy, now it does show the debug logs.  Thanks to your help, I've figured out how to handle everything I've implemented so far, and I am confident that I can manage most of what's to come.

I've just found ${DEFAULT-VALUE}, which I will use in many of my descriptions.

![image](https://user-images.githubusercontent.com/4984827/236727247-74c8524f-02a1-47c5-9e0c-3034cf5be8cb.png)
